prober/icmp.go

```diff
@@ -143,21 +153,23 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 		if module.ICMP.DontFragment {
 			// If the user has set the don't fragment option we cannot use unprivileged
 			// sockets as it is not possible to set IP header level options.
-			icmpConn, err := net.ListenPacket("ip4:icmp", srcIP.String())
+			netConn, err := net.ListenPacket("ip4:icmp", srcIP.String())
 			if err != nil {
 				level.Error(logger).Log("msg", "Error listening to socket", "err", err)
 				return
 			}
 
-			rc, err := ipv4.NewRawConn(icmpConn)
+			level.Info(logger).Log("msg", "Creating RawConn")
+			rawConn, err = ipv4.NewRawConn(netConn)
 			if err != nil {
 				level.Error(logger).Log("msg", "Error creating raw connection", "err", err)
 				return
 			}
-			socket = &v4Conn{c: rc, df: true}
+			if err := rawConn.SetControlMessage(ipv4.FlagTTL, true); err != nil {
+				level.Debug(logger).Log("msg", "Failed to set Control Message for retrieving TTL", "err", err)
+			}
+			defer rawConn.Close()
```

Comment:
Shouldn't any defer be right after it is instantiated? That reduces the risk of it being missed.

Reply:
There is just one place it could fail with an open connection before ipv4.NewRawConn(). I added the extra Close() as I guess multiple Close() will not break anything. Anyway, I moved all defer to before SetControlMessage.